QAchat/python/QAchat.ipynb

```diff
@@ -0,0 +1,428 @@
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n",
+      "\n",
+      "=========================SUMMARY=========================\n",
+      "pipelineA-001 compared to pipelineB-001:\n",
+      "90.00% (-10.00%) 'saysThankYou' score\t(0 improvements, 4 regressions)\n",
+      "48.50% (-00.50%) 'factuality'   score\t(7 improvements, 6 regressions)\n",
+      "28.58% (-12.05%) 'levenshtein'  score\t(2 improvements, 15 regressions)\n",
+      "85.00% (-10.00%) 'saysBadWords' score\t(0 improvements, 4 regressions)\n",
+      "\n",
+      "See results for all experiments in QAchatbot at https://www.braintrustdata.com/app/braintrustdata.com/p/QAchatbot\n",
+      "See results for pipelineA-001 at https://www.braintrustdata.com/app/braintrustdata.com/p/QAchatbot/pipelineA-001\n"
+     ]
+    }
+   ],
+   "source": [
+    "# 4. Run against different prompts and pipelines\n",
+    "\n",
+    "# A very simple pipeline that uses 1 fact as context\n",
+    "def llmPipelineA(question):\n",
+    "    # Get a relevant fact from the user context dataset\n",
+    "    relevant = collection.query(\n",
+    "        query_texts=[question],\n",
+    "        n_results=1\n",
+    "    )\n",
+    "    relevant_text = ','.join(relevant[\"documents\"][0])\n",
+    "    prompt = \"\"\"\n",
+    "        You are an assistant called BT. Help the user. Do not say you are an AI language model. Follow up with -BT.\n",
+    "        Relevant information: {relevant}\n",
+    "        Question: {question}\n",
+    "        Answer:\n",
+    "        \"\"\".format(question=question, relevant=relevant_text)\n",
+    "    messages = [{\"role\": \"system\", \"content\": prompt}]\n",
+    "    response = openai.ChatCompletion.create(\n",
```

Comment:
I think we should use async everywhere -- we want to set a good example for building these tutorials performantly